### PR TITLE
Manual-cherry-pick-for-1.24: kubeadm: apply registry.k8s.io changes

### PR DESCRIPTION
**Pull request by neolit123**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

manual, kubeadm scoped backport of:
https://github.com/kubernetes/kubernetes/pull/109938
https://github.com/kubernetes/kubernetes/pull/110343

#### Which issue(s) this PR fixes:

xref https://github.com/kubernetes/kubeadm/issues/2671

#### Special notes for your reviewer:

it is essentially a feature backport, as discussed with k8s-infra and sig-release

https://kubernetes.slack.com/archives/C2C40FMNF/p1666723622759119
https://github.com/kubernetes/kubeadm/issues/2671#issuecomment-1290991218

#### Does this PR introduce a user-facing change?

```release-note
kubeadm: use the image registry registry.k8s.io instead of k8s.gcr.io for new clusters. During upgrade, migrate users to registry.k8s.io if they were using the default of k8s.gcr.io.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: build/dependencies.yaml
Comment (SergeyKanzhelev) [inline build/dependencies.yaml:172]: I believe this needs to be moved as well to `registry.k8s.io/pause`

Comment (SergeyKanzhelev) [reply build/dependencies.yaml:172]: And also one above that declares dependencies for the Makefile only.

This is how it looks in master now: https://github.com/kubernetes/kubernetes/blob/819cadca66a7dccfa5c0772605fc90c667555c97/build/dependencies.yaml#L147-L154

Comment (SergeyKanzhelev) [reply build/dependencies.yaml:172]: Maybe it's best to approve this one: https://github.com/kubernetes/kubernetes/pull/113403/files that covers it

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go
Comment (SataQiu) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go:33]: Emm... This doesn't seem to be in the same order we had before.
We always put the kubeadm import last.

File: build/dependencies.yaml
Comment (SataQiu) [reply build/dependencies.yaml:172]: Maybe it seems better to use a separate PR to make these changes.

**Comment by neolit123** [comment]
@SataQiu 
i've updated all 3 prs as they had this minor problem:
https://github.com/kubernetes/kubernetes/pull/113395#discussion_r1007900415

https://github.com/kubernetes/kubernetes/pull/113395
https://github.com/kubernetes/kubernetes/pull/113393
https://github.com/kubernetes/kubernetes/pull/113388

**Comment by SataQiu** [comment]
/lgtm
/approve
for kubeadm

File: build/dependencies.yaml
Comment (SataQiu) [inline build/dependencies.yaml:154]: Sorry, but I'm a little confused here.
The value in the following `match` field is still `k8s.gcr.io`, which looks inconsistent with the `name` description.
Do we really need to make this change in this PR?

Comment (neolit123) [reply build/dependencies.yaml:154]: right, this should have only matched against kubeadm paths

Comment (neolit123) [reply build/dependencies.yaml]: updated this part

**Comment by neolit123** [comment]
@dims can you please approve this too for /build

**Comment by neolit123** [comment]
@jackfrancis @CecileRobertMichon 
we need someone to address these test errors from the capz team. can you help?

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/113395/pull-kubernetes-e2e-capz-conformance/1593066526968123392
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/113395/pull-kubernetes-e2e-capz-conformance/1593066526968123392/build-log.txt

> nregistry.k8s.io/kube-controller-manager-amd64:v1.24.9-rc.0.15_7bfb91a2e90bd9" is not a valid repository/tag: invalid reference format
Error parsing reference: "k8s.gcr.io/kube-proxy-amd64:v1.24.9-rc.0.15_7bfb91a2e90bd9\nregistry.k8s.io/kube-proxy-amd64:v1.24.9-rc.0.15_7bfb91a2e90bd9" is not a valid repository/tag: invalid reference format
Error parsing reference: "k8s.gcr.io/kube-scheduler-amd64:v1.24.9-rc.0.15_7bfb91a2e90bd9\nregistry.k8s.io/kube-scheduler-amd64:v1.24.9-rc.0.15_7bfb91a2e90bd9" is not a valid repository/tag: invalid reference format
The push refers to repository

**Comment by jackfrancis** [comment]
@neolit123 without going into detail I'm fairly certain a recent rev to a new capz release is responsible for this. This reverts that change:

https://github.com/kubernetes/test-infra/pull/28050

**Comment by jackfrancis** [comment]
@neolit123 I have been running the capz conformance tests locally against a (fixed) capz branch, and it doesn't work against this PR, something is going wrong during the source build flow during image tagging:

```
+++ [1118 19:42:34] Deleting conformance image k8s.gcr.io/conformance-amd64:v1.24.8-rc.0.18_f7058b0a588731
+++ [1118 19:42:34] Docker builds done
+++ [1118 19:42:34] Building tarball: server linux-amd64
+++ [1118 19:43:20] Building tarball: final
+++ [1118 19:43:20] Waiting on test tarballs
+++ [1118 19:43:20] Starting tarball: test linux-amd64
+++ [1118 19:43:51] Building tarball: test portable
Error parsing reference: "k8s.gcr.io/kube-apiserver-amd64:v1.24.8-rc.0.18_f7058b0a588731\nregistry.k8s.io/kube-apiserver-amd64:v1.24.8-rc.0.18_f7058b0a588731" is not a valid repository/tag: invalid reference format
Error parsing reference: "k8s.gcr.io/kube-controller-manager-amd64:v1.24.8-rc.0.18_f7058b0a588731\nregistry.k8s.io/kube-controller-manager-amd64:v1.24.8-rc.0.18_f7058b0a588731" is not a valid repository/tag: invalid reference format
Error parsing reference: "k8s.gcr.io/kube-proxy-amd64:v1.24.8-rc.0.18_f7058b0a588731\nregistry.k8s.io/kube-proxy-amd64:v1.24.8-rc.0.18_f7058b0a588731" is not a valid repository/tag: invalid reference format
Error parsing reference: "k8s.gcr.io/kube-scheduler-amd64:v1.24.8-rc.0.18_f7058b0a588731\nregistry.k8s.io/kube-scheduler-amd64:v1.24.8-rc.0.18_f7058b0a588731" is not a valid repository/tag: invalid reference format
```

**Comment by neolit123** [comment]
seems like there is an \n in the image name/tag.
(stating the obvious i guess)

**Comment by jackfrancis** [comment]
Is it worth auditing the changes in this PR against the corresponding foo in `release-1.25`? (I'm not getting this repo against 1.25 branch, which already has the gcr.k8s.io -> registry.k8s.io stuff, right?)

**Comment by neolit123** [comment]
you could try. it is not entirely clear to me where this replacement and the \n is coming from.

**Comment by CecileRobertMichon** [comment]
@jackfrancis looking at where in the script the error is happening, it looks like it might be https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e552aab3a1a22378ea51aa81fe8d55cc915b5dd4/scripts/ci-build-kubernetes.sh#L95 

Potentially that line is no longer resilient when k8s builds two different images with two different registries as it was before? And (completely hypothesizing here) maybe 1.25 isn't seeing the same symptoms because k8s started building with the new registry only starting with k8s 1.25.

**Comment by jackfrancis** [comment]
@neolit123 if I apply the change to `build/lib/release.sh` to this branch that was a part of @dims's original PR it seems to unblock things:

https://github.com/kubernetes/kubernetes/commit/50bea1dad89930ad565526910aadc314b9e9f38b#diff-8b2df901800c5898df862f8fb42532fefd3c151ca2797d0cdeb8fb9cd250573a

Does that make any sense to you?

**Comment by neolit123** [comment]
> Does that make any sense to you?

i can push a second commit in the PR with this.

**Comment by cpanato** [comment]
@dims can you review this one as well as you did the others? thanks

**Comment by neolit123** [comment]
@jackfrancis the capz job passed after the release.sh change. thanks

/retest

**Comment by neolit123** [comment]
wondering if we have to backport more from the original diff to make these GCE jobs to pass.
@BenTheElder could you help please?

/retest

**Comment by neolit123** [comment]
@jackfrancis NVM, looks like it's still failing.

**Comment by jackfrancis** [comment]
@CecileRobertMichon is correct, the tarballs from the kube release flow now have two images identified with them:

```
$ docker load --input _output/release-images/amd64/kube-apiserver.tar
Loaded image: k8s.gcr.io/kube-apiserver-amd64:v1.24.8-rc.0.18_f7058b0a588731
Loaded image: registry.k8s.io/kube-apiserver-amd64:v1.24.8-rc.0.18_f7058b0a588731
```

@neolit123 is that a concern, or can we confidently update our CI to always pull the first image (e.g., `head -n 1`) in order to adapt our existing capz scripts to accommodate this change?

**Comment by neolit123** [comment]
> the tarballs from the kube release flow now have two images identified with them:
> is that a concern

i think it is, @dims @BenTheElder do you know how to resolve this in the diff of this PR?

**Comment by dims** [comment]
@neolit123 let's wait for Ben to chime in.

**Comment by jackfrancis** [comment]
capz tests are passing but they aren't running against the correct build of k8s. I was able to test this out successfully locally:

```
$ k get nodes -o wide
NAME                                   STATUS   ROLES           AGE   VERSION                          INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
capz-conf-nqzq8y-control-plane-n4lh6   Ready    control-plane   22m   v1.24.8-rc.0.18+a15ecbe053e82b   10.0.0.4      <none>        Ubuntu 18.04.6 LTS   5.4.0-1091-azure   containerd://1.6.2
capz-conf-nqzq8y-md-0-7k2nz            Ready    <none>          19m   v1.24.8-rc.0.18+a15ecbe053e82b   10.1.0.4      <none>        Ubuntu 18.04.6 LTS   5.4.0-1091-azure   containerd://1.6.2
capz-conf-nqzq8y-md-0-t4rf2            Ready    <none>          19m   v1.24.8-rc.0.18+a15ecbe053e82b   10.1.0.5      <none>        Ubuntu 18.04.6 LTS   5.4.0-1091-azure   containerd://1.6.2
```

But this job seems to be running against "latest-stable". The kubelet version from the latest job:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/113395/pull-kubernetes-e2e-capz-conformance/1595243679646224384/artifacts/clusters/capz-conf-77q7ia/machines/capz-conf-77q7ia-control-plane-bp2nv/kubelet-version.txt

... matches the current latest 24 build:

https://storage.googleapis.com/k8s-release-dev/ci/latest-1.24.txt

(The value from the 2nd URL will obviously change as commits land in `release-1.24` but as of now it's returning `v1.24.9-rc.0.11+d818028a185189`.)

cc @CecileRobertMichon

**Comment by neolit123** [comment]
@dims ben's slack status says he is on vacation. what parts do we need to fix the gce jobs?

@jackfrancis i see. so this is a latest-x.txt marker problem right now and it will be fixed later?

@pacoxu all kubeadm 1.24 periodics are now broken during init. i wonder how that happened. 
https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-1-24
i will have a look today.

**Comment by pacoxu** [comment]
> https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-1-24

This should be green after this PR is merged.
1.24 is trying to pull `k8s.gcr.io` image while we have changed the kinder to use `registry.k8s.io` after https://github.com/kubernetes/kubeadm/pull/2783 was merged.

1.22-to-1.23 upgrade ci is green after https://github.com/kubernetes/kubernetes/pull/113393 for v1.23 is merged.

**Comment by CecileRobertMichon** [comment]
> But this job seems to be running against "latest-stable". The kubelet version from the latest job:
... matches the current latest 24 build

@jackfrancis are you sure that's not a symptom of the way that template is setup to first run kubeadm init/join with latest of that branch and _then_ replace container images to use the images built from the PR? That kubelet-version.txt file probably gets written once by kubeadm when it initially bootstraps the node.

What version are the nodes running right before we start running conformance?

**Comment by jackfrancis** [comment]
@CecileRobertMichon build output from that job shows it's running against the wrong 1.24 version (the "latest" 1.24 at the time of the test):

- https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/113395/pull-kubernetes-e2e-capz-conformance/1595243679646224384/build-log.txt

The commit it actually ran against should have resolved the version to `v1.24.8-rc.0.18+a15ecbe053e82b`.

Running another local test now, still not repro'ing.

In-progress output on another job kicked off from this PR suggests it is once again not choosing the right commit:

```
Login Succeeded
/home/prow/go/src/k8s.io/kubernetes /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure
/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   138  100   138    0     0   4600      0 --:--:-- --:--:-- --:--:--  4600

100    30  100    30    0     0    250      0 --:--:-- --:--:-- --:--:--   250
Error response from daemon: manifest for capzci.azurecr.io/kube-apiserver:v1.24.9-rc.0.16_f043d48880e746 not found: manifest unknown: manifest tagged by "v1.24.9-rc.0.16_f043d48880e746" is not found
```

(should be `0082fddeafe5f4`, thus `v1.24.8-rc.0.19_0082fddeafe5f4` <-- this is what my local repro is doing against the latest commit in this PR branch)

**Comment by jackfrancis** [comment]
I'm suspicious that when we run this flow in capz that the `KUBE_GIT_VERSION` env var isn't correct. I have a PR to help us debug that further:

- https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2849

Once we get that cherry-picked to capz @ `release-1.6` I can re-run the capz conformance job against this PR and get more info.

**Comment by jackfrancis** [comment]
Confirmed that the `KUBE_GIT_VERSION` environment variable is being set to "latest 1.24 build" rather than the commit from the source fork commit in this PR.

This is what capz is doing to get that input:

```
    source "${KUBE_ROOT}/hack/lib/version.sh"
    pushd "${KUBE_ROOT}" && kube::version::get_version_vars && popd
    : "${KUBE_GIT_VERSION:?Environment variable empty or not defined.}"
    export KUBE_GIT_VERSION
```

Has capz always been relying upon `KUBE_GIT_VERSION` incorrectly? Should our test foo rely upon another method to get the commit SHA to produce something like `v1.24.8-rc.0.19+0082fddeafe5f4`? Again, this flow works locally when I set up everything like test-infra expect in my local dev environment, but perhaps there's something more interesting that happens in real time when prow sets up the infra and checks out the k/k and capz repos to run this test against PRs.

**Review by cpanato — APPROVED** [review]
/triage accepted
/priority important-soon
/lgtm

**Comment by puerco** [comment]
Thank you so much!

**Pull request merged**
